feat: Add OverRule for learning Boolean rules to describe support/overlap

moberst opened the pull request:
This PR implements the OverRule method from [1], for learning boolean rule sets to characterize (a) the support of the data, and (b) overlap between treatment and control groups. Code is adapted (with some simplifications) from https://github.com/clinicalml/overlap-code, under the MIT License.

Currently implemented as a refutation method via both the `CausalModel` API as well as the functional API
```
refute = model.refute_estimate(identified_estimand, estimate, method_name='assess_overlap')
print(refute)  # Returns the rules learned and statistics regarding coverage of existing samples
```
An example notebook is included in `docs/source/example_notebooks/dowhy_refuter_assess_overlap.ipynb`.

Opening a PR to facilitate discussion on how best to integrate, @amit-sharma @emrekiciman 

[1] Oberst, M., Johansson, F., Wei, D., Gao, T., Brat, G., Sontag, D., & Varshney, K. (2020). Characterization of
Overlap in Observational Studies. In S. Chiappa & R. Calandra (Eds.), Proceedings of the Twenty Third International
Conference on Artificial Intelligence and Statistics (Vol. 108, pp. 788–798). PMLR. https://arxiv.org/abs/1907.04138

**Additional Dependencies**
This change adds `cvxpy` as a dependency, and uses `xgboost` for the default propensity score model.  The latter could be changed if adding `xgboost` as a main dependency (currently just a dev dependency) is not desirable.

emrekiciman commented:
This is very nice, @moberst!  I just glanced through the sample notebook, and I like how this works as a refuter, and the list of rules it returns.  

I'm looking forward to digging deeper into the PR soon.

emrekiciman commented:
@moberst it looks like there's a failure in the automated checks.

@darthtrevino do you know why the git clone command might be failing in the build?

darthtrevino commented:
@emrekiciman The new "can install from source" check isn't pulling from @moberst 's fork. We can ignore that check for this PR, but let's make sure CI passes. I'll dig into correcting that GH action.

moberst commented:
@darthtrevino If there's anything I can do on my end to help that action / check work, please just let me know.

One question on dependency management, which I mentioned above but wanted to re-highlight:  I use `xgboost` as a dependency, which seems to work fine when running tests, but I noticed that in the `pyproject.toml` file this is in listed under the `dev` dependencies group.

With that in mind, should I be modifying `pyproject.toml` to require `xgboost` under `[tool.poetry.dependencies]`, instead of under `[tool.poetry.group.dev.dependencies]`, where it currently sits?

darthtrevino commented:
@moberst I don't see a need yet to promote xgboost into a full-blown dependency. Can you merge in the latest changes from main?

moberst commented:
No problem - I removed the part where I use `xgboost` in this PR, to avoid requiring it as a dependency.  I also noticed that one of the notebook tests failed because `seaborn` was not available, so I reworked to use `matplotlib` instead in the notebook.

darthtrevino commented:
Seaborn is already a documentation dependency, let's move that to the dev section. 

WRT xgboost, we could make it an optional dependency, but iirc it was only used in tests (I'm not at my computer atm). Are you using it in an estimator or some other module in the source folder?

moberst commented on dowhy/causal_refuters/assess_overlap_overrule.py:
```diff
@@ -0,0 +1,310 @@
+
+        if prop_estimator is None:
+            param_grid = {"max_depth": [4, 8], "min_samples_leaf": [0.01, 0.1], "n_estimators": [200]}
+            estimator = RandomForestClassifier(criterion="log_loss", random_state=0)
```
@darthtrevino To your question about `xgboost`, I was using `XGBClassifier` as a default propensity score model, but have swapped that for `RandomForestClassifier`, which is no problem from my perspective

amit-sharma reviewed (commented):
Thanks for adding this functionality @moberst. This will be valuable contribution to the community. 

The code and docs quality is excellent. I added some inline comments. Here are some high-level thoughts: 
1. It will be good to add example code in the notebook that connects the method directly to the effect inference task. I suggested LaLonde in an inline comment, but even a toy dataset will help. 
2. In addition to refutation, do you think we can use Overrule to identify the individual units on which to estimate causal effect? In `dowhy.estimate_effect` function, we have a parameter called `target_units` where users can provide a dataframe (or a lambda function) to provide the individuals for which the causal effect should be computed. In addition to the boolean rules, if Overrule can return a subset of the dataframe or a filtering function, then its output can be used directly as `target_units` to determine the individuals on which causal estimation should be done.  Does this make sense? I may have misunderstood, so please correct me if not. 

I also notice that the PR depends on a new dependency, "cvxpy". @darthtrevino do you think this dependency is okay to include as a requirement? It looks light-weight enough, but good to check.

emrekiciman commented on dowhy/causal_refuters/assess_overlap.py:
```diff
@@ -0,0 +1,124 @@
+        :raises NotImplementedError: Will raise this error if show_progress_bar=True
+        """
+        if show_progress_bar:
+            raise NotImplementedError("No progress bar is available for OverRule")
```
Do we want to raise a NotImplementedError here?  Given that the progress bar is secondary functionality, I think we should we just print a warning that it's not implemented but otherwise continue with the rest of the functionality.

moberst replied:
Changed to a warning, good call

emrekiciman commented on dowhy/causal_refuters/overrule/BCS/beam_search.py:
```diff
@@ -0,0 +1,201 @@
+
+class PricingInstance:
+    """
+    Instance of the pricing problem.
```
What is the pricing problem?  I wonder if there's a name for this class that would be more intuitive in a causal inference context.  Alternatively, it would be sufficient to point to a formal description of the pricing problem

moberst replied:
Added a citation - this doesn't mean much in a causal inference context, it's an implementation detail of how the optimization problem is solved.

emrekiciman reviewed (commented):
Agreed with @amit-sharma that this code looks to be overall of high quality.  And the assessment of overlap and support is super important.  Thank you for your contribution!

I have a couple general questions as well:
* Given a new or arbitrary unit attribute, is there a way to query the generated ruleset to evaluate whether the unit falls within the overlap/support region?  I didn't see one, but maybe I missed it?  This seems like a useful building block for visualizations and for use cases where we are learning models and then applying them to new data.
* Is there any special consideration that must be given when interpreting support and overlap when using CATE estimators?
* when we are estimating direct effect of a treatment (as opposed to total effect) we also condition on the mediators.  Are there any special considerations when interpreting the support and overlap then?
* Is evaluation of support something that is more broadly applicable, beyond backdoor adjustment methods?
* Is this overlap and support evaluation something that could be adapted to support the GCM modeling as well? (just interested in getting your thoughts here --- not for this PR)

I don't think any of these are blockers, though perhaps they are items we'd want to make sure are documented, or maybe some turn into follow up PRs

moberst commented on docs/source/example_notebooks/dowhy_refuter_assess_overlap.ipynb:
```diff
@@ -0,0 +1,508 @@
+    "\n",
+    "## Motivation\n",
+    "\n",
+    "Even if all other causal assumptions hold, we can only estimate causal effects for individuals with covariates $X$ that satisfy two conditions: \n",
```
Agreed - fixed in the latest version

moberst commented on docs/source/example_notebooks/dowhy_refuter_assess_overlap.ipynb:
```diff
@@ -0,0 +1,508 @@
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Rules cover 50.0% of all samples\n",
```
The interpretation for support and overlap rules is the same here, since they are both DNF.  A data point is eligible if *any* of the support rules and *any* of the overlap rules apply.

Note that, in part to make it easier to review the code & standardize the output, I've restricted to learning only DNF rules in this PR - the example from the paper you're referring to uses CNF rules for the output.  

I added clarifications on this in the output itself, and in the notebook.

moberst commented on docs/source/example_notebooks/dowhy_refuter_assess_overlap.ipynb:
```diff
@@ -0,0 +1,508 @@
+    "* `support_config`: A `SupportConfig` (see below), which controls optimization parameters\n",
+    "\n",
+    "* `overlap_config`: An `OverlapConfig` (see below), which controls optimization parameters"
+   ]
```
I've added a Lalonde example to the notebook.  However, the experimental sample from Lalonde isn't a great test bed by itself, since it's an experiment - in fact, I show in the notebook that overlap holds for all samples :) 

I've included something like what we show in the paper - if you combined the experimental samples from Lalonde with observational controls from another dataset, OverRule does a good job at identifying a resulting lack of overlap.

moberst commented:
Thank you both for the feedback!  Addressing general comments here:

# Adding an example with real data
(from @amit-sharma)
> 1. It will be good to add example code in the notebook that connects the method directly to the effect inference task. I suggested LaLonde in an inline comment, but even a toy dataset will help.

I've added a Lalonde example to the notebook. However, the experimental sample from Lalonde isn't a great test bed by itself, since it's an experiment - in fact, I show in the notebook that overlap holds for all samples :)

I've included something like what we show in the paper - if you combined the experimental samples from Lalonde with observational controls from another dataset, OverRule does a good job at identifying a resulting lack of overlap.

# Querying the ruleset to classify points as inside/outside the overlap/support region

(from @amit-sharma)
> 2. In addition to refutation, do you think we can use Overrule to identify the individual units on which to estimate causal effect? In `dowhy.estimate_effect` function, we have a parameter called `target_units` where users can provide a dataframe (or a lambda function) to provide the individuals for which the causal effect should be computed. In addition to the boolean rules, if Overrule can return a subset of the dataframe or a filtering function, then its output can be used directly as `target_units` to determine the individuals on which causal estimation should be done.  Does this make sense? I may have misunderstood, so please correct me if not.

(from @emrekiciman)
> Given a new or arbitrary unit attribute, is there a way to query the generated ruleset to evaluate whether the unit falls within the overlap/support region? I didn't see one, but maybe I missed it? This seems like a useful building block for visualizations and for use cases where we are learning models and then applying them to new data.

Yes!  I've added an explicit function to the refuter that does this, and I give an example of applying it at the end of the notebook.

# Other points

(from @emrekiciman)
> Is there any special consideration that must be given when interpreting support and overlap when using CATE estimators?

No special considerations - if overlap does not hold in a region, then CATE estimation is effectively relying on extrapolation to estimate effects.

> when we are estimating direct effect of a treatment (as opposed to total effect) we also condition on the mediators. Are there any special considerations when interpreting the support and overlap then?

Good question!  Not sure I have a great answer off the top of my head.  As it stands, the question of overlap estimation here is meant for contexts where we perform backdoor adjustment as opposed to performing mediation analysis.

> Is evaluation of support something that is more broadly applicable, beyond backdoor adjustment methods?

It certainly could be applied in other contexts - note that the refuter takes an argument to only learn support rules as well, in cases where that might be the main concern.

> Is this overlap and support evaluation something that could be adapted to support the GCM modeling as well? (just interested in getting your thoughts here --- not for this PR)

I think it certainly could be - for learning individual causal mechanisms, it might be worth characterizing which parts of the conditional distribution have support in the data.

emrekiciman commented:
@darthtrevino @moberst do either of you know why there might be a merge conflict in the poetry.lock file and how to resolve it?  (see the "content_hash" entry on the last line of poetry.lock)

moberst commented:
> @darthtrevino @moberst do either of you know why there might be a merge conflict in the poetry.lock file and how to resolve it? (see the "content_hash" entry on the last line of poetry.lock)

My understanding is that `poetry.lock` updates that hash every time the lock file is updated.  I've now done the following, which I believe is the correct course of action here (@darthtrevino please feel free to correct me)
* Merge `main` into my branch, which includes several updates to `poetry.lock`
* Rebuild the lockfile using `poetry lock --no-update`, which refreshes the lock file to be consistent with `pyproject.toml` but does not update locked versions

emrekiciman reviewed (approved):
Thanks for addressing my comments and questions, @moberst  This will be very useful.

emrekiciman commented:
@andresmor-ms , do you have any comments on this PR from the engineering perspective?

@amit-sharma did @moberst's updates resolve your questions too?

darthtrevino commented:
@emrekiciman with a merge in a lockfile we should merge in the latest from main to clear the conflict, and then rebuild & commit the lockfile. As long as the CI/CD pipelines pass that should be safe. Lockfiles change pretty frequently.

amit-sharma reviewed (approved):
Thanks @moberst Really appreciate the lalonde/PSID example which brings out the utility of overrule refuter. Approving.

amit-sharma commented:
@all-contributors please add @moberst for code.
